holoviews/core/util.py

```diff
@@ -809,36 +809,80 @@ def find_range(values, soft_range=[]):
             return (None, None)
 
 
-def max_range(ranges):
+def is_finite(value):
+    """
+    Safe check whether a value is finite, only None and NaN values are
+    considered non-finite and allows checking all types not restricted
+    to numeric types.
+    """
+    if value is None:
+        return False
+    try:
+        return np.isfinite(value)
+    except:
+        return True
+
+
+def max_range(ranges, combined=True):
     """
     Computes the maximal lower and upper bounds from a list bounds.
     """
     try:
         with warnings.catch_warnings():
             warnings.filterwarnings('ignore', r'All-NaN (slice|axis) encountered')
-            values = [r for r in ranges for v in r if v is not None]
+            values = [tuple(np.NaN if v is None else v for v in r) for r in ranges]
             if pd and all(isinstance(v, pd.Timestamp) for r in values for v in r):
                 values = [(v1.to_datetime64(), v2.to_datetime64()) for v1, v2 in values]
             arr = np.array(values)
-            if arr.dtype.kind in 'OSU':
-                arr = list(python2sort([v for v in arr.flat if not is_nan(v) and v is not None]))
+            if not len(arr):
+                return np.NaN, np.NaN
+            elif arr.dtype.kind in 'OSU':
+                arr = list(python2sort([v for r in values for v in r if not is_nan(v) and v is not None]))
                 return arr[0], arr[-1]
-            if arr.dtype.kind in 'M':
-                return arr[:, 0].min(), arr[:, 1].max()
-            return (np.nanmin(arr[:, 0]), np.nanmax(arr[:, 1]))
+            elif arr.dtype.kind in 'M':
+                return (arr.min(), arr.max()) if combined else (arr[:, 0].min(), arr[:, 1].min())
+            if combined:
+                return (np.nanmin(arr), np.nanmax(arr))
+            else:
+                return (np.nanmin(arr[:, 0]), np.nanmax(arr[:, 1]))
     except:
         return (np.NaN, np.NaN)
 
 
-def dimension_range(lower, upper, dimension):
+def range_pad(lower, upper, padding=None, log=False):
+    """
+    Pads the range by a fraction of the interval
```

Comment:
I didn't spot any unit tests explicitly for these utilities. I know they should be tested indirectly by the more comprehensive unit tests in this PR but a couple of utility unit tests might be useful as examples of the expected input/output...

Reply:
Agreed, should be straightforward.